WIP: Refactor proxy code now that proxy ports are shared

jrajahalme opened the pull request:
Simplify proxy code now that each redirect does not have a unique proxy port.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8552)
<!-- Reviewable:end -->

joestringer commented on pkg/proxy/proxy.go:
```diff
@@ -103,6 +103,29 @@ type Proxy struct {
+		defer p.mutex.Unlock()
+
+		// Check that Envoy is started when needed
+		if rTypes&^policy.RedirectTypeAgentMask != 0 {
```
These kinds of bitwise checks (in this case logically `rTypes.RequiresExternalProxy()`) could be factored out for easier readability and unit-tested to ensure they have the intended behaviour.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -611,6 +610,26 @@ func NewL4Policy(revision uint64) *L4Policy {
 }
 
+// HasRedirects returns true if the L4Policy needs any L7 proxy redirects.
+func (l4 *L4Policy) HasRedirects() bool {
```
Since all of these new functions regarding redirects are interacting directly with the `l4.redirects` field, you should be able to define these functions against the `RedirectType` type and directly compose the `RedirectType` into `L4Policy` to inherit them for free. Then, since the functions are all straightforward bitwise manipulations we can unit-test the `RedirectType` to ensure that the correct low-level behaviour is occurring.

joestringer commented on pkg/proxy/proxy.go:
```diff
@@ -642,30 +638,13 @@ func (p *Proxy) removeRedirect(id string, wg *completion.WaitGroup) (err error,
-			time.Sleep(portReuseDelay)
-
-			proxyPortsMutex.Lock()
-			err := p.releaseProxyPort(listenerName)
```
I take it that proxy port lifetime is now tied to the listener?

jrajahalme replied:
This has been the case for a while now, redirects have shared proxy ports via shared listeners since about the time TPROXY was merged in.

joestringer reviewed (approved):
I don't fully follow the internal proxy changes (commits 8/9), but this appears to be mostly just shuffling code around. I made a couple of minor suggestions for unit-testability below, but LGTM assuming it passes the CI.

Do we want to wait some amount of time after 1.6 branch to minimize potential backport conflicts?

ianvernon commented on daemon/daemon.go:
```diff
@@ -188,17 +188,17 @@ func (d *Daemon) Datapath() datapath.Datapath {
 
 // UpdateProxyRedirect updates the redirect rules in the proxy for a particular
 // endpoint using the provided L4 filter. Returns the allocated proxy port
```
Update the comment.

jrajahalme committed:
policy: Cache L4Policy redirect info

Precompute a mask of all redirect types needed by an L4Policy.

Replace HasRedirect() function with the new 'hasIngressRedirect()'.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Refactor StartProxySupport

Proxy package already takes in some of the parameters directly from
the option package. Refactor StartProxySupport() so that the
configuration parameters are passed via options rather than function
parameters. This helps separating starting the Envoy process from the
creation of a redirect. Introduce StartEnvoy() for the same purpose.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Introduce StartProxies()

daemon.StartProxies() can be used to start the external proxy processes
independently of the creation of a redirect.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Shift the call to start Envoy from create redirect to policy computation

Start the host proxy Envoy process when required by any policy
applicable to a local endpoint. This allows the proxy port to be known
when the mapstate entries are created.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Use FinalizeList

Use FinalizeList to manage multiple finalize functions instead of open coding it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Remove unused arguments

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Fix comments

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Factor out listener start from creating a redirect

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Remove waitgroups that are no longer needed.

Proxy updates are either synchronous, or managed via xDS so the WaitGroups can be removed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Create listeners prior to configuring redirects

Create listeners before redirects are created. Listeners are now never
stopped once they have been started.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Avoid scanning L4Filters with no redirects

Since L4Policy keeps track of the needed proxy types, we can check if
we need to go through the L4Filters at all.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Do not create "implementation" for Envoy redirects

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Create MapState entries in one pass

Remove the 2nd pass to patch up proxy ports by correctly populating
proxy ports on the 1st pass.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Simplify Redirect interface

Redirect.Close() does not need the revert function, so remove it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Fix comment

UpdateProxyRedirect no longer returns the proxy port.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Move 'rules' to the redirect implementation

Move 'rules' from proxy/redirect to the redirect implementation. This
removes duplication for the DNS proxy, as now the 'rules' is changed
under the control of the implementation. The passed in rules map is
immutable, so there is no need to deep copy it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
